phalcon/acl/adapter/memory.zep

```diff
@@ -780,6 +749,88 @@ class Memory extends Adapter
 		return haveAccess == Acl::ALLOW;
 	}
 
+	/**
+	 * Check whether a role is allowed to access an action from a resource
+	 */
+	protected function _isAllowed(string roleName, string resourceName, string access) -> string | boolean
+    {
+        var accessList, accessKey,checkRoleToInherit, checkRoleToInherits, usedRoleToInherits, usedRoleToInherit;
+
+		let accessList = this->_access;
+
+        let accessKey = roleName . "!" . resourceName . "!" . access;
+
+		/**
+		 * Check if there is a direct combination for role-resource-access
+		 */
+		if isset accessList[accessKey] {
+			return accessKey;
+		}
+		/**
+         * Check if there is a direct combination for role-*-*
+         */
+        let accessKey = roleName . "!" . resourceName . "!*";
+        if isset accessList[accessKey] {
+            return accessKey;
+        }
+        /**
+         * Check if there is a direct combination for role-*-*
+         */
+        let accessKey = roleName . "!*!*";
+        if isset accessList[accessKey] {
+            return accessKey;
+        }
+        /**
+         * Deep check if the role to inherit is valid
+         */
+        if isset this->_roleInherits[roleName] {
+            let checkRoleToInherits = [];
+            for usedRoleToInherit in this->_roleInherits[roleName] {
+                let checkRoleToInherits[] = usedRoleToInherit;
+            }
+            let usedRoleToInherits = [];
+            while !empty checkRoleToInherits {
+                let checkRoleToInherit = array_shift(checkRoleToInherits);
```

Comment:
Not sure also about this, isn't checkRoleToInherits[0] and unset if necessary enough?

Reply:
this is  a stack, need popping。

Reply:
What's the situation now?

Reply:
@Jurigag This creates the risk of an infinite loop！
```
let checkRoleToInheritsIndex = 0;
                while !empty checkRoleToInherits {
                    if empty(checkRoleToInherits[checkRoleToInheritsIndex]) {
                        let checkRoleToInheritsIndex ++;
                        continue;
                    }
                    let checkRoleToInherit = checkRoleToInherits[checkRoleToInheritsIndex];
                    unset checkRoleToInherits[checkRoleToInheritsIndex];
                    let checkRoleToInheritsIndex ++;
```